### ISSUE TITLE
[ECS] [request]: Support Cross Service Cross Task Definition Dependency

### ISSUE THREAD

**Issue by e-moshaya** [aws/containers-roadmap#482]
**Tell us about your request**
What do you want us to build?
Support Cross Service Cross Task Definition Dependency

**Which service(s) is this request for?**
Fargate, ECS

Based on aws documentation https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_definitions.html, 

"Your entire application stack does not need to exist on a single task definition, and in most cases it should not. Your application can span multiple task definitions by combining related containers into their own task definitions, each representing a single component"

This means that it is best practice to separate each component/service into its own task definition. However, cross task definition is not yet supported. Only container dependencies within a single task definition is supported. This needs extending to support cross service.

**Comment by coultn** [comment]
Can you explain more about your use case?  Do you want to have services that start up and shut down in a certain order?  Or something else?  The more detail you can provide, the better we can evaluate and prioritize.  
Thanks for your feedback!

**Comment by e-power** [comment]
+1 for this.  

As you said @coultn we would like to prioritise startup order of different services which can run into issues if a certain service starts before the other (currently random). This has come up basically as a result of starting and stopping the ECS cluster overnight for cost saving reasons.

**Comment by e-power** [comment]
+1 for this.

As you said @coultn we would like to prioritise startup order of different services which can run into issues if a certain service starts before the other (currently random). This has come up basically as a result of starting and stopping the ECS cluster overnight for cost saving reasons.

**Comment by kuchibha** [comment]
Here is a possible scenario: I would like to run task definitions in a dependent fashion and on a scheduled basis. Ex- Run Task1 (on success) -> Run Task2 (on success) -> Run Task3

Possible options to accomplish this:
a. 1 task definition with multiple container definitions.
Each container definition with its own "Command" and "container dependency"- for ex- container 2 should run after SUCCESS of container1. At-least one container needs to be marked ESSENTIAL, which means failure of that container will stop the task. This prohibits me from retrying the failed task ( in a set of container dependencies)

b. Create 1 task definition for each container. How do I set task-definition **dependency**? For ex- Task-Definition-2 should run after Task-Definition-1.

Either way, I would need a scope for retry upon failure of any one task.

**Comment by snegostup** [comment]
Need this to run Apollo GraphQL Federated Schema.
In short, there is a Gateway that polls child schemas (Fargate services) on startup to compile a single schema. The Gateway must spawn after all child services are up. Otherwise it won't contain parts of schema that were spawned after it.

**Comment by kamatama41** [comment]
We have a daemon service of [datadog-agent](https://docs.datadoghq.com/integrations/amazon_ecs/) on an ECS EC2 cluster. As the agent service collects metrics and logs from other ECS services, the service should start before (and stop after) any other services when an EC2 instance is initializing (and draining). It would be achieved if we can define the dependency between services.

**Comment by mooreniemi** [comment]
K8s has [initContainers](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/#init-containers-in-use) pattern. Looks like this can be done with [Container Depdencies](https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_ContainerDependency.html) in ECS.

**Comment by afdecastro879** [comment]
HI all!
Is there any findings around this ticket?
Our use case is we want to run a setup task before we spin up our tasks in ECS, the problem is, this setup task should be run only once even if we want to spin up more than one task for our application. I've checked the Container Dependencies in ECS and this doesn't work as it will run the `initial container` for each instance of our app.
As far as I understand the purpose of the ticket is how to make sure independent tasks can depend on each other at the service level and not at the container level

**Comment by sjohnsonsf** [comment]
+1 vote for this feature. Use case is a Rails app, running a task to completes ### db:migrate first, then running the app server after migrations have completed.

**Comment by jas550** [comment]
+1 vote for this feature. Use case is a php app, first the db migration task needs to run once and then php app should execute after migrations have completed successfully.

**Comment by eeepmb** [comment]
+1 for the same reason as @afdecastro879 

> HI all! Is there any findings around this ticket? Our use case is we want to run a setup task before we spin up our tasks in ECS, the problem is, this setup task should be run only once even if we want to spin up more than one task for our application. I've checked the Container Dependencies in ECS and this doesn't work as it will run the `initial container` for each instance of our app. As far as I understand the purpose of the ticket is how to make sure independent tasks can depend on each other at the service level and not at the container level

**Comment by pbudzon** [comment]
This is also important when you add in the fact that ECS has a limit on task definition size (65536 bytes). This means that even if you tried to put in your application within one task (to use the `DependsOn` from task definition), it quickly becomes impossible because you hit the size limit.

**Comment by TreehouseFalcon** [comment]
+1, this issue has been open for about 5 years now and I'm shocked it isn't natively supported in ECS. For my use case, I have a postgres and redis database that need to be deployed first and then my dependent services afterwards.

**Comment by steven-geo** [comment]
+1 Here. I have a service container that other containers need to connect to, but are not directly part of that core service. Using DependsOn seems to only work internally in the same service/task, not to other services. Adding an option to Depend on a service would be great, with the same condition options as DependsOn.

**Comment by alandreasyans** [comment]
+1

**Comment by Parth2404** [comment]
+1
